pan-domain-auth-verification/src/main/scala/com/gu/pandomainauth/Settings.scala

```diff
@@ -1,66 +1,105 @@
-  } catch {
-    case NonFatal(e) =>
-      Left(SettingsDownloadFailure(e))
+  /**
+   * @param settingsFileKey the name of the file that contains the private settings for the given domain
+   */
+  class Loader(s3BucketLoader: S3BucketLoader, settingsFileKey: String) {
+
+    def loadAndParseSettingsMap(): SettingsResult[Map[String, String]] = fetchSettings().flatMap(extractSettings)
+
+    private def fetchSettings(): SettingsResult[String] = try {
+      Right(IOUtils.toString(s3BucketLoader.inputStreamFetching(settingsFileKey)))
+    } catch { case NonFatal(e) => Left(SettingsDownloadFailure(e)) }
   }
 
-  private[pandomainauth] def extractSettings(settingsBody: String): Either[SettingsFailure, Map[String, String]] = try {
+  private[pandomainauth] def extractSettings(settingsBody: String): SettingsResult[Map[String, String]] = try {
     val props = new Properties()
     props.load(new ByteArrayInputStream(settingsBody.getBytes("UTF-8")))
-
     Right(props.asScala.toMap)
   } catch {
     case NonFatal(e) =>
       Left(SettingsParseFailure(e))
   }
 
-  def logError(failure: SettingsFailure, logger: Logger) = failure match {
-    case SettingsDownloadFailure(cause) =>
-      logger.error("Unable to download public key", cause)
-
-    case SettingsParseFailure(cause) =>
-      logger.error("Unable to parse public key", cause)
+  class Refresher[A](
+    loader: Settings.Loader,
+    settingsParser: Map[String, String] => SettingsResult[A],
+    scheduler: ScheduledExecutorService = Executors.newScheduledThreadPool(1)
+  ) {
+    // This is deliberately designed to throw an exception during construction if we cannot immediately read the settings
+    private val store: AtomicReference[A] = new AtomicReference(
+      loadAndParseSettings().fold(fail => throw fail.asThrowable(), identity)
+    )
 
-    case PublicKeyFormatFailure =>
-      logger.error("Public key does not match expected format")
-
-    case PublicKeyNotFoundFailure =>
-      logger.error("Public key not found in settings file")
-  }
+    private val logger = LoggerFactory.getLogger(getClass)
 
-  def errorToThrowable(failure: SettingsFailure): Throwable = failure match {
-    case SettingsDownloadFailure(cause) =>
-      new IllegalStateException("Unable to download public key", cause)
+    def start(interval: Int): Unit = scheduler.scheduleAtFixedRate(() => refresh(), 0, interval, MINUTES)
 
-    case SettingsParseFailure(cause) =>
-      new IllegalStateException("Unable to parse public key", cause)
+    def loadAndParseSettings(): SettingsResult[A] =
+      loader.loadAndParseSettingsMap().flatMap(settingsParser)
 
-    case PublicKeyFormatFailure =>
-      new IllegalStateException("Public key does not match expected format")
+    private def refresh(): Unit = loadAndParseSettings() match {
+      case Right(newSettings) =>
+        // logger.debug(s"Updated pan-domain settings for $domain")
+        val oldSettings = store.getAndSet(newSettings)
```

Comment:
it doesn't look like we're doing anything with `oldSettings` here?

(and should the commented-out logger line above be removed?)

Reply:
Good point - addressed with 412c1d6f4efa42c68f6f25cec73a74d850b604ad!